File: solutions/102. Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        #soln 1 - recursion
        ans = []
        if not root:
            return root
        
        def helper(node, level):
            if len(ans) == level:
                ans.append([])
            ans[level].append(node.val)
            
            if node.left:
                helper(node.left, level+1)
            if node.right:
                helper(node.right, level+1)
        helper(root, 0)
        return ans
    
        # #soln 0 - first attempt w deque
        # if not root: return root
        # ans = []
        # queue = collections.deque([root])
        # while queue:
        #     size = len(queue)
        #     temp = []
        #     for i in range(size):
        #         node = queue.popleft()
        #         temp.append(node.val)
        #         if node.left: queue.append(node.left)
        #         if node.right: queue.append(node.right)
        #     ans.append(temp)
        # return ans

File: solutions/102. Binary Tree Level Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Soln 0 - see similar problem 314. Binary Tree Vertical Order Traversal
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans = {};
        if (!root) return ans;
        
        std::deque<std::pair<TreeNode*, int>> deq;
        deq.push_back(std::make_pair(root, 0));
        while (!deq.empty()) {
            const auto pr = deq.front();
            const auto& node = pr.first;
            int row = pr.second;
            deq.pop_front();
​
            if (row >= ans.size()) {
                ans.push_back(vector<int> {node->val});
            } else {
                ans[row].push_back(node->val);
            }
            if (node->left) {
                deq.push_back(std::make_pair(node->left, row+1));
            } 
            if (node->right) {
                deq.push_back(std::make_pair(node->right, row+1));
            }
        }
        
        return ans;
    }
};
